drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+			rx_frame.id  = (idr >> CAN_XLNX_FIFO_IDR_IDL_OFFSET) &
+				CAN_XLNX_FIFO_IDR_IDL_MASK;
+			rx_frame.id |= ((idr >> CAN_XLNX_FIFO_IDR_IDH_OFFSET) &
+					CAN_XLNX_FIFO_IDR_IDH_MASK) << 18;
+
+			/* RTR -> acquire from [RTR] */
+			if (idr & CAN_XLNX_FIFO_IDR_RTR) {
+				rx_frame.rtr = CAN_REMOTEREQUEST;
+			} else {
+				rx_frame.rtr = CAN_DATAFRAME;
+			}
+		} else {
+			rx_frame.id_type = CAN_STANDARD_IDENTIFIER;
+
+			/* Standard ID in IDH only */
+			rx_frame.id  = (idr >> CAN_XLNX_FIFO_IDR_IDH_OFFSET) &
+				CAN_XLNX_FIFO_IDR_IDH_MASK;
+
+			/* RTR -> acquire from [SRRRTR] */
+			if (idr & CAN_XLNX_FIFO_IDR_SRRRTR) {
+				rx_frame.rtr = CAN_REMOTEREQUEST;
+			} else {
+				rx_frame.rtr = CAN_DATAFRAME;
+			}
+		}
+
+		/* Frame payload */
+		rx_frame.data_32[0] = ntohl(dw1);
+		rx_frame.data_32[1] = ntohl(dw2);
+
+		/*
+		 * The frame has been read from the RX FIFO unconditionally,
+		 * now check if it matches and registered RX filter.
+		 */
+		if (dev_data->filter_usage_mask) {
+			for (i = 0; i < CAN_XLNX_NUM_ACCEPTANCE_FILTERS; i++) {
+				/*
+				 * Skip the current filter entry if:
+				 * - it is unused (corresp. bit not set in filter_usage_mask)
+				 * - the filter's ID/mask don't match the RX frame's ID.
+				 */
+				if (!(dev_data->filter_usage_mask & BIT(i))) {
+					continue;
+				}
+
+				if (!can_utils_filter_match(&rx_frame,
+							    &dev_data->rx_filters[i].filter)) {
+					continue;
+				}
+
+				dev_data->rx_filters[i].cb(dev, &rx_frame,
+							   dev_data->rx_filters[i].cb_arg);
+			}
+		} else {
+			LOG_DBG("%s discarded received frame with ID 0x%X as no "
+				"RX filters are registered", dev->name, rx_frame.id);
+		}
+
+		/* Get updated interrupt status register */
+		isr = sys_read32(base_addr + CAN_XLNX_ISR_OFFSET);
+
+		/* Check for RX FIFO undeflow error */
+		if (isr & CAN_XLNX_IRQ_RX_UNDERFLOW) {
+			sys_write32(CAN_XLNX_IRQ_RX_UNDERFLOW,
+				    base_addr + CAN_XLNX_ICR_OFFSET);
+			LOG_ERR("%s read from RX FIFO caused an undeflow error",
+				dev->name);
+			break;
+		}
+	} while (isr & CAN_XLNX_IRQ_RX_NOT_EMPTY);
+}
+
+/**
+ * @brief TX FIFO transmission done handler function
+ * @internal
+ *
+ * Called from within the ISR if the TX OK interrupt bit is set in
+ * the Interrupt Status Register, this function executes each sent
+ * message's transmission done handler. This can either be the com-
+ * bination of a callback function pointer plus its parameter pointer,
+ * or a pointer to a semaphore on whose acquisition the can_xlnx_send
+ * function is currently blocking.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_handle_tx(const struct device *dev)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+	struct can_xlnx_dev_data *dev_data = dev->data;
+	struct can_xlnx_tx_done_handler *handler;
+
+	/* Clear the TX OK interrupt bit */
+	sys_write32(CAN_XLNX_IRQ_MESSAGE_TX, base_addr + CAN_XLNX_ICR_OFFSET);
+
+	/* Process all pending TX done handlers */
```

Comment:
Why process all pending TX done "handlers" here? Only the callbacks for the frames actually transmitted are to be called.